dhis-2/dhis-services/dhis-service-program-rule/src/main/java/org/hisp/dhis/programrule/hibernate/HibernateProgramRuleStore.java

```diff
@@ -110,17 +112,52 @@ public List<ProgramRule> getProgramRulesByActionTypes(
   @Override
   public List<ProgramRule> getProgramRulesByActionTypes(
       Program program, Set<ProgramRuleActionType> types, String programStageUid) {
-    final String hql =
-        "SELECT distinct pr FROM ProgramRule pr JOIN FETCH pr.programRuleActions pra "
-            + "LEFT JOIN FETCH pr.programStage ps "
-            + "WHERE pr.program = :programId AND pra.programRuleActionType IN ( :implementableTypes ) "
-            + "AND (pr.programStage IS NULL OR ps.uid = :programStageUid )";
-
-    return getQuery(hql)
-        .setParameter("programId", program)
-        .setParameter("implementableTypes", types)
-        .setParameter("programStageUid", programStageUid)
-        .getResultList();
+    String hql =
+        """
+                SELECT distinct pr FROM ProgramRule pr JOIN FETCH pr.programRuleActions pra
```

Comment:
Could we use native sql here?
It would be easier to understand what is happening and we would have more control on this query.
We don't need the `FETCH` keyword here as we are getting the actions in the next query.
In this query we just need to filter out program rules with no actions of certain types

Reply:
converting this into native should not be an issue.
But I do think we need join with `ProgramRuleAction` to filter rules based on `ProgramRuleActionType`.

Reply:
We need the join, we do not need the fetch part

Reply:
fixed